### PR TITLE
Add a new parameter when creating a ENet server …

### PR DESCRIPTION
**Pull request by santiagoalvarez**
So that the server itself doesn't propagate to all peers when a new peer connects/disconnects.

This is useful when wanting to run a pure server

### CODE REVIEW

**Comment by mhilbrunner** [comment]
I don't get how not notifying other peers when peers connect or disconnect makes the server *pure*. Could you explain what the use case for this is?

**Comment by santiagoalvarez** [comment]
My current use case is having a "matchmaking" server where there is no need for clients to know other clients connected. I'll be handling this in a different manner.
As in, no client needs to know of other clients connected to this specific server since it's only a middle-man before the match itself.

**Comment by mhilbrunner** [comment]
cc @Faless

**Comment by Faless** [comment]
> My current use case is having a "matchmaking" server where there is no need for clients to know other clients connected. I'll be handling this in a different manner.

Why not using use a simple TCP socket then?

**Comment by santiagoalvarez** [comment]
I would like to be able to still use ENet so I can mix reliable and unreliable calls, and use the rpc* calls.
I believe it also makes sense to be able to be able to turn off this extra behavior.

**Comment by Faless** [comment]
As it is it will also break `rpc_id`

**Comment by santiagoalvarez** [comment]
Ok, can you explain a bit more why it breaks rpc_id (since it was not my intention)?
Is it only because the "clients" will no longer know the id of other clients?
The matchmaking was just an example, its not the only use case.
I would also need this in a match server and turning off any "peer-to-peer" like behavior would be nice.

To me this is a simple change that allows some extra behavior that is built on top of ENet to be turned off.

**Comment by Faless** [comment]
> Is it only because the "clients" will no longer know the id of other clients?

yes, RPCs will not be sent by the clients if they don't have the `peer_id` in the hashmap

> The matchmaking was just an example, its not the only use case.

Like?

> I would also need this in a match server and turning off any "peer-to-peer" like behavior would be nice.

I'm just trying to understand why you need the MultiplayerAPI without the peers knowing who is connected. What is the problem you are trying to solve? If you are worried about one peer flooding the others we could add an option to disable replaying of messages from the server. But I see no real reason on why to skip the peer notification part.

**Comment by santiagoalvarez** [comment]
Adding an option to disbale replaying messages is a great idea.
In essence my use case is the basic FPS Multiplayer game where clients are untrusted and every action is handled by the server only and the only trusted data also comes only from the server, but also that not every client in itself is a player (think of someone that just joinned to watch, the peers shouldn't need to know someone connected in such a case). With that in mind, having peers notified when other peers joins brings no benefit since I will never rpc_id to any other id other than 1 from clients.
While being able to both udp and tcp at the same time brings some value (not every data needs to be udp and some info I would like to go over tcp for reliability purposes).
To summarize I would like to only have packets go server <--> client and never client 1 <--> server <-->  client 2 (I would categorise the connection/disconnection notification as the latter). It will be wasted resources as whatever information it carries will go unused and I can see a possible issue if someone uses a malicious client that keeps connecting and disconnecting trying to waste server resources (removing some extra steps would not eliminate the threat of course, but would use less resources).

**Comment by mhilbrunner** [comment]
Closing this, as the change as proposed breaks networking features.

After talking with Faless it may indeed be worthwhile to be able to disable the p2p/forwarding/relaying features, and he has an idea how to do it in the future. No timeline though, sorry.
